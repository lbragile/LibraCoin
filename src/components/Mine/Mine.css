.block {
  width: 25%;
  padding: 10px;
  border-radius: 10px;
  position: relative;
}

.block label {
  margin-bottom: 0;
}

.bordered-background {
  filter: brightness(90%);
  border: 1px dashed black;
  border-radius: 10px;
  padding: 1em;
}
