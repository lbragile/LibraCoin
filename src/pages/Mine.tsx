import React from "react";

import NavMenu from "../components/NavMenu/NavMenu";
import PreviewBlock from "../components/Block/PreviewBlock";
import TransactionItems from "../components/Transaction/TransactionItems";

export default function Mine(): JSX.Element {
  return (
    <div className="container-fluid">
      <NavMenu />
      <TransactionItems />
      <PreviewBlock />
    </div>
  );
}
