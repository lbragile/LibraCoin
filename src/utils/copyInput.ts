export function copyKey(
  e: React.FocusEvent<HTMLTextAreaElement | HTMLInputElement>,
  setCopied: (arg: boolean[]) => void,
  type?: "public" | "private"
): void {
  e.target.select();
  e.target.setSelectionRange(0, 1e6);
  document.execCommand("copy");

  // wallet page, copying keys
  if (type) {
    const isPublic = type === "public";
    setCopied([isPublic, !isPublic && !e.target.value.includes("â—¦")]);
  } else {
    setCopied([true]);
  }
}
