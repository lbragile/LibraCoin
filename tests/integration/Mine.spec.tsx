/**
 * @group integration
 */

import React from "react";
import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Mine from "../../src/pages/Mine";
import * as ConversionUtil from "../../src/utils/conversion";
import { customRender } from "../utils/testUtils";

const { initialState } = global;

it("starts off invalid, and after mining it stays invalid (cannot be added to blockchain)", async () => {
  const solution = "000z4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";
  const target = "000y4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";

  Date.now = jest.fn().mockReturnValueOnce(12345);
  jest.spyOn(ConversionUtil, "randomHash").mockReturnValue("random");
  jest.spyOn(ConversionUtil, "digestMessage").mockResolvedValueOnce(target).mockResolvedValueOnce(solution);

  const { asFragment } = customRender(<Mine />);

  expect(screen.getByRole("form", { name: /Block Statistics/i })).toHaveFormValues({
    nonce: 0,
    header: 0,
    target: "",
    solution: ""
  });

  expect(screen.getByRole("form", { name: /Block Form/i })).toHaveFormValues({
    ...initialState.preview,
    valid: undefined,
    transactions: undefined
  });

  expect(screen.queryByRole("button", { name: /Add Block/i })).not.toBeInTheDocument();
  expect(screen.getByRole("form", { name: /Block Form/i })).toHaveClass("invalid-block");
  expect(screen.getByRole("textbox", { name: /Block Solution/i })).toHaveClass("invalid-solution");

  const verifiedTrans = screen.getAllByRole("form", { name: /Transaction Information/i });
  const selectedTrans = verifiedTrans.filter((trans) => trans.classList.contains("selected"));
  expect(verifiedTrans).toHaveLength(initialState.verifiedTrans.length);
  expect(selectedTrans).toHaveLength(initialState.selectedTrans.length);

  expect(asFragment()).toMatchSnapshot();

  userEvent.click(screen.getByRole("button", { name: /Block Mine/i }));

  // once mining is complete
  await waitFor(() => expect(screen.getByRole("status")).toHaveClass("invisible"));
  expect(screen.getByRole("form", { name: /Block Form/i })).toHaveClass("invalid-block");
  expect(screen.getByRole("textbox", { name: /Block Solution/i })).toHaveClass("invalid-solution");
  expect(screen.queryByRole("button", { name: /Add Block/i })).not.toBeInTheDocument();

  const newVerifiedTrans = screen.getAllByRole("form", { name: /Transaction Information/i });
  const newSelectedTrans = newVerifiedTrans.filter((trans) => trans.classList.contains("selected"));
  expect(newVerifiedTrans).toStrictEqual(verifiedTrans);
  expect(newSelectedTrans).toStrictEqual(selectedTrans);
});

it("starts off invalid, but after mining turns valid, and is added to blockchain", async () => {
  const solution = "000a4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";
  const target = "000z4fda363405b2796986a63e8cedde080e1f29ed774f5f93bd97c42b9a96fc0";

  Date.now = jest.fn().mockReturnValueOnce(12345).mockReturnValueOnce(67890); // first time right after mining, second time in addition of block
  jest.spyOn(ConversionUtil, "randomHash").mockReturnValue("random");
  jest.spyOn(ConversionUtil, "digestMessage").mockResolvedValueOnce(target).mockResolvedValueOnce(solution);

  const { asFragment } = customRender(<Mine />);

  expect(screen.getByRole("form", { name: /Block Statistics/i })).toHaveFormValues({
    nonce: 0,
    header: 0,
    target: "",
    solution: ""
  });

  expect(screen.getByRole("form", { name: /Block Form/i })).toHaveFormValues({
    ...initialState.preview,
    valid: undefined,
    transactions: undefined
  });

  expect(screen.queryByRole("button", { name: /Add Block/i })).not.toBeInTheDocument();
  expect(screen.getByRole("form", { name: /Block Form/i })).toHaveClass("invalid-block");
  expect(screen.getByRole("textbox", { name: /Block Solution/i })).toHaveClass("invalid-solution");

  const verifiedTrans = screen.getAllByRole("form", { name: /Transaction Information/i });
  const selectedTrans = verifiedTrans.filter((trans) => trans.classList.contains("selected"));
  expect(verifiedTrans).toHaveLength(initialState.verifiedTrans.length);
  expect(selectedTrans).toHaveLength(initialState.selectedTrans.length);

  expect(asFragment()).toMatchSnapshot();

  userEvent.click(screen.getByRole("button", { name: /Block Mine/i }));

  // once mining is complete
  await waitFor(() => expect(screen.getByRole("status")).toHaveClass("invisible"));
  expect(screen.getByRole("form", { name: /Block Form/i })).toHaveClass("valid-block");
  expect(screen.getByRole("textbox", { name: /Block Solution/i })).toHaveClass("valid-solution");
  expect(screen.getByRole("button", { name: /Add Block/i })).toBeInTheDocument();

  const afterMiningVerifiedTrans = screen.getAllByRole("form", { name: /Transaction Information/i });
  const afterMiningSelectedTrans = afterMiningVerifiedTrans.filter((trans) => trans.classList.contains("selected"));
  expect(afterMiningVerifiedTrans).toStrictEqual(verifiedTrans);
  expect(afterMiningSelectedTrans).toStrictEqual(selectedTrans);

  userEvent.click(screen.getByRole("button", { name: /Add Block/i }));

  // once block is added to blockchain
  await waitFor(() => expect(screen.getByRole("form", { name: /Block Form/i })).toHaveClass("invalid-block"));
  expect(screen.getByRole("textbox", { name: /Block Solution/i })).toHaveClass("invalid-solution");

  expect(screen.getAllByRole("form", { name: /Transaction Information/i })).toHaveLength(
    verifiedTrans.length - initialState.selectedTrans.length
  );

  const afterAddVerifiedTrans = screen.getAllByRole("form", { name: /Transaction Information/i });
  const afterAddSelectedTrans = afterAddVerifiedTrans.filter((trans) => trans.classList.contains("selected"));
  expect(afterAddSelectedTrans).toEqual([]);
  expect(afterAddVerifiedTrans).toStrictEqual(
    verifiedTrans.filter((trans) => trans.classList.contains("not-selected"))
  );

  await waitFor(() =>
    expect(screen.getByRole("form", { name: /Block Form/i })).toHaveFormValues({
      index: initialState.preview.index + 1,
      timestamp: 67890,
      prevHash: solution,
      currHash: "",
      merkleRoot: ""
    })
  );
});
