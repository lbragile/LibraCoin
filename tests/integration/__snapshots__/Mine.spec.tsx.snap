// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`starts off invalid, and after mining it stays invalid (cannot be added to blockchain) 1`] = `
<DocumentFragment>
  .c0 {
  min-height: 18rem;
}

.c1 {
  margin: 0.5rem;
  padding: 0.5em;
  border-radius: 5px;
  border: 2px solid transparent;
  min-width: 14rem;
  background-color: #fff;
}

.c1:hover {
  cursor: pointer;
  border: 2px solid #090;
}

.c3 {
  margin: 0.5rem;
  padding: 0.5em;
  border-radius: 5px;
  border: 2px solid transparent;
  min-width: 14rem;
  background-color: #dfd;
}

.c3:hover {
  cursor: pointer;
  border: 2px solid #090;
}

.c2.c2.c2 {
  font-weight: bold;
}

<div
    class="container-fluid"
  >
    <h3
      aria-label="Title"
      class="font-weight-bold"
    >
      Verified Transactions
    </h3>
    <div
      class="c0 row flex-nowrap overflow-auto bg-dark mx-1 px-2 rounded"
    >
      <form
        aria-label="Transaction Information"
        class="c1"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="B"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            First Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="1.23"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AB1.23"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information Selected"
        class="c3"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="C"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Second Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="456.78"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AC456.78"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="c1"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="D"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Third Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="999.99"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AD999.99"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="c1"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="E"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Forth Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="987.65"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AE987.65"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="c1"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="F"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Fifth Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="1.01"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AF1.01"
          />
        </div>
      </form>
    </div>
  </div>
  .c1.c1.c1 {
  font-weight: bold;
}

.c2 {
  color: #f00;
}

.c3 {
  background-color: #ff9a9a;
}

.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

<div
    class="container-fluid row d-flex justify-content-center mx-auto my-3"
  >
    <div
      class="w-100 mb-2 text-center"
    >
      <h3
        class="font-weight-bold w-25 mx-auto"
      >
        Merkle Tree Visualization
      </h3>
      <p
        class="my-3 border w-75 mx-auto"
      >
        Please select a verified transaction from above...
      </p>
    </div>
    <form
      aria-label="Block Statistics"
      class="c0 col-12 col-lg-5 mx-3 mb-2 mb-lg-0 bg-dark px-2 pb-2 rounded"
    >
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Nonce
          </span>
        </div>
        <input
          aria-label="Block Nonce"
          class="form-control"
          disabled=""
          name="nonce"
          type="number"
          value="0"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Header
          </span>
        </div>
        <input
          aria-label="Block Header"
          class="form-control"
          disabled=""
          name="header"
          type="number"
          value="0"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Target
          </span>
        </div>
        <input
          aria-label="Block Target"
          class="text-truncate form-control"
          name="target"
          readonly=""
          type="text"
          value=""
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Sol'n
          </span>
        </div>
        <input
          aria-label="Block Solution Invalid"
          class="c2 text-truncate form-control"
          name="solution"
          readonly=""
          type="text"
          value=""
        />
      </div>
      <button
        aria-label="Block Mine"
        class="btn-block p-1 btn btn-primary"
        type="button"
      >
        <h4
          class="my-0 row justify-content-end align-items-center flex-nowrap"
        >
          <span
            class="col-10 pl-5"
          >
            Mine
          </span>
          <div
            class="mr-3 invisible spinner-border"
            role="status"
          >
            <span
              class="sr-only"
            >
              Loading...
            </span>
          </div>
        </h4>
      </button>
    </form>
    <form
      aria-label="Block Form Invalid"
      class="c3 col-12 col-lg-5 p-2 rounded"
    >
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Index
          </span>
        </div>
        <input
          aria-label="Block Index"
          class="form-control"
          disabled=""
          name="index"
          type="number"
          value="3"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Timestamp
          </span>
        </div>
        <input
          aria-label="Block Timestamp"
          class="form-control"
          disabled=""
          name="timestamp"
          type="number"
          value="1619913600000"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Previous #
          </span>
        </div>
        <input
          aria-label="Block PrevHash"
          class="text-truncate form-control"
          name="prevHash"
          readonly=""
          type="text"
          value="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Current #
          </span>
        </div>
        <input
          aria-label="Block CurrHash"
          class="text-truncate form-control"
          name="currHash"
          readonly=""
          type="text"
          value="BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB"
        />
      </div>
      <div
        class="mt-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Merkle #
          </span>
        </div>
        <input
          aria-label="Block Merkle"
          class="text-truncate form-control"
          name="merkleRoot"
          readonly=""
          type="text"
          value="previewMerkle"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`starts off invalid, but after mining turns valid, and is added to blockchain 1`] = `
<DocumentFragment>
  .c0 {
  min-height: 18rem;
}

.c1 {
  margin: 0.5rem;
  padding: 0.5em;
  border-radius: 5px;
  border: 2px solid transparent;
  min-width: 14rem;
  background-color: #fff;
}

.c1:hover {
  cursor: pointer;
  border: 2px solid #090;
}

.c3 {
  margin: 0.5rem;
  padding: 0.5em;
  border-radius: 5px;
  border: 2px solid transparent;
  min-width: 14rem;
  background-color: #dfd;
}

.c3:hover {
  cursor: pointer;
  border: 2px solid #090;
}

.c2.c2.c2 {
  font-weight: bold;
}

<div
    class="container-fluid"
  >
    <h3
      aria-label="Title"
      class="font-weight-bold"
    >
      Verified Transactions
    </h3>
    <div
      class="c0 row flex-nowrap overflow-auto bg-dark mx-1 px-2 rounded"
    >
      <form
        aria-label="Transaction Information"
        class="c1"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="B"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            First Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="1.23"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AB1.23"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information Selected"
        class="c3"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="C"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Second Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="456.78"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AC456.78"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="c1"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="D"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Third Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="999.99"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AD999.99"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="c1"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="E"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Forth Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="987.65"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AE987.65"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="c1"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="F"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Fifth Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="1.01"
          />
          <div
            class="input-group-append"
          >
            <span
              class="c2 input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="c2 input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AF1.01"
          />
        </div>
      </form>
    </div>
  </div>
  .c1.c1.c1 {
  font-weight: bold;
}

.c2 {
  color: #f00;
}

.c3 {
  background-color: #ff9a9a;
}

.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

<div
    class="container-fluid row d-flex justify-content-center mx-auto my-3"
  >
    <div
      class="w-100 mb-2 text-center"
    >
      <h3
        class="font-weight-bold w-25 mx-auto"
      >
        Merkle Tree Visualization
      </h3>
      <p
        class="my-3 border w-75 mx-auto"
      >
        Please select a verified transaction from above...
      </p>
    </div>
    <form
      aria-label="Block Statistics"
      class="c0 col-12 col-lg-5 mx-3 mb-2 mb-lg-0 bg-dark px-2 pb-2 rounded"
    >
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Nonce
          </span>
        </div>
        <input
          aria-label="Block Nonce"
          class="form-control"
          disabled=""
          name="nonce"
          type="number"
          value="0"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Header
          </span>
        </div>
        <input
          aria-label="Block Header"
          class="form-control"
          disabled=""
          name="header"
          type="number"
          value="0"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Target
          </span>
        </div>
        <input
          aria-label="Block Target"
          class="text-truncate form-control"
          name="target"
          readonly=""
          type="text"
          value=""
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Sol'n
          </span>
        </div>
        <input
          aria-label="Block Solution Invalid"
          class="c2 text-truncate form-control"
          name="solution"
          readonly=""
          type="text"
          value=""
        />
      </div>
      <button
        aria-label="Block Mine"
        class="btn-block p-1 btn btn-primary"
        type="button"
      >
        <h4
          class="my-0 row justify-content-end align-items-center flex-nowrap"
        >
          <span
            class="col-10 pl-5"
          >
            Mine
          </span>
          <div
            class="mr-3 invisible spinner-border"
            role="status"
          >
            <span
              class="sr-only"
            >
              Loading...
            </span>
          </div>
        </h4>
      </button>
    </form>
    <form
      aria-label="Block Form Invalid"
      class="c3 col-12 col-lg-5 p-2 rounded"
    >
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Index
          </span>
        </div>
        <input
          aria-label="Block Index"
          class="form-control"
          disabled=""
          name="index"
          type="number"
          value="3"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Timestamp
          </span>
        </div>
        <input
          aria-label="Block Timestamp"
          class="form-control"
          disabled=""
          name="timestamp"
          type="number"
          value="1619913600000"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Previous #
          </span>
        </div>
        <input
          aria-label="Block PrevHash"
          class="text-truncate form-control"
          name="prevHash"
          readonly=""
          type="text"
          value="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
        />
      </div>
      <div
        class="my-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Current #
          </span>
        </div>
        <input
          aria-label="Block CurrHash"
          class="text-truncate form-control"
          name="currHash"
          readonly=""
          type="text"
          value="BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB"
        />
      </div>
      <div
        class="mt-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c1 input-group-text"
          >
            Merkle #
          </span>
        </div>
        <input
          aria-label="Block Merkle"
          class="text-truncate form-control"
          name="merkleRoot"
          readonly=""
          type="text"
          value="previewMerkle"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
